### DISCUSSION TITLE
CommonAPI C++ and vsomeip-repository

### DISCUSSION THREAD

**Discussion by nachtfuchs** [COVESA/vsomeip#535] [Q&A] [answered]
I do not understand the idea of "CommonAPI C++". Especially, I do not understand its linking to vsomeip. Why is the [wiki ](https://github.com/COVESA/vsomeip/wiki)in the vsomeip linking to the CommonAPI? Is the CommonAPI an alternative to vsomeip? Is it a predecessor? Can anybody help me and shed some light into it?

**Comment by BrunoLSilvaCTW** [comment, marked as answer]
Hi @nachtfuchs,

CommonAPI  is an abstraction layer over Inter-Process Communication (IPC) protocols, specifically, SOME/IP and DBus. It enables you to write code without having to worry about which binding you use under the hood.

vsomeip is a standalone implementation of the SOME/IP protocol, and it's what CommonAPI uses as its SOME/IP binding.

I hope this answers your questions.

**Reply by nachtfuchs** [reply]
Hi @BrunoLSilvaCTW,
Can you elaborate on "which binding you use under the hood"? If I want to use the SOME/IP protocol, do I still need to install the vsomeip-library as the binding? Or does it come in a package in CommonAPI?
What would be an alternative binding to vsomeip for CommonAPI for SOME/IP?

**Reply by BrunoLSilvaCTW** [reply]
> Can you elaborate on "which binding you use under the hood"?

Either DBus or SOME/IP (vsomeip).

> If I want to use the SOME/IP protocol, do I still need to install the vsomeip-library as the binding? Or does it come in a package in CommonAPI?

You would need to install vsomeip separately. 

> What would be an alternative binding to vsomeip for CommonAPI for SOME/IP?

There is no alternative to vsomeip for CommonAPI SOME/IP.

**Reply by nachtfuchs** [reply]
Got it. I googled DBus only after your message. So CommonAPI makes life easier when someone wants to run SOME/IP and D-Bus. If someone wants to use only SOME/IP, then CommonAPI seems to be of no immediate benefit. What would you say?

**Reply by nachtfuchs** [reply]
If there is no such a documentation that describes the overall idea of CommonAPI, vsomeip, D-Bus, then I think it would be worth mentioning it. This content is something I would usually read in a SW architecture documentation. Example is [arc42](https://arc42.org/). Any chance there is a SW architecture documentation for the mentioned topic?

**Reply by BrunoLSilvaCTW** [reply]
The closest thing is the [CommonAPICppSpecification](https://github.com/COVESA/capicxx-core-tools/blob/7bf7b2371fda4cb43beb495e1fc52feb0b480f59/docx/CommonAPICppSpecification), which should give you a good understanding of CommonAPI's purpose. ~~Some of the old releases included this as a [pdf](https://github.com/COVESA/capicxx-core-tools/releases/download/3.1.12/CommonAPICppUserGuide.pdf), but~~ You should be able to generate a pdf of this yourself.

*Edit: The releases included the user guide, not the specification.*

**Comment by fcmonteiro** [comment]
Hi @nachtfuchs,
We have some [CommonAPI-Examples here](https://github.com/COVESA/capicxx-core-tools/tree/master/CommonAPI-Examples), maybe it will help you understand how to use it.
The repository is a little bit outdated, but it will be updated soon.
There is already a PR there for the update.

**Reply by nachtfuchs** [reply]
Thank you for the link. I read about the HelloWorld example in the wiki and quickly browsed the mentioned examples. What I believe to have understood is, that CommonAPI uses a different structure to create and request services than vsomeip. This is where I stopped digging deeper in the understanding, because it is not a use-case for me to use one more layer on top of SOME/IP.
